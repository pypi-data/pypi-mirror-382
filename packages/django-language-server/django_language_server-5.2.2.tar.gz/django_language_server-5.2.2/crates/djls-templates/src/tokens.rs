use djls_source::Span;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum TagDelimiter {
    Block,
    Variable,
    Comment,
}

impl TagDelimiter {
    pub const CHAR_OPEN: char = '{';
    pub const LENGTH: usize = 2;
    pub const LENGTH_U32: u32 = 2;

    #[must_use]
    pub fn from_input(input: &str) -> Option<Self> {
        let bytes = input.as_bytes();

        if bytes.len() < Self::LENGTH {
            return None;
        }

        if bytes[0] != Self::CHAR_OPEN as u8 {
            return None;
        }

        match bytes[1] {
            b'%' => Some(Self::Block),
            b'{' => Some(Self::Variable),
            b'#' => Some(Self::Comment),
            _ => None,
        }
    }

    #[must_use]
    pub fn opener(self) -> &'static str {
        match self {
            Self::Block => "{%",
            Self::Variable => "{{",
            Self::Comment => "{#",
        }
    }

    #[must_use]
    pub fn closer(self) -> &'static str {
        match self {
            Self::Block => "%}",
            Self::Variable => "}}",
            Self::Comment => "#}",
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Token {
    Block { content: String, span: Span },
    Comment { content: String, span: Span },
    Error { content: String, span: Span },
    Eof,
    Newline { span: Span },
    Text { content: String, span: Span },
    Variable { content: String, span: Span },
    Whitespace { span: Span },
}

impl Token {
    /// Get the content text for content-bearing tokens
    #[must_use]
    pub fn content(&self) -> String {
        match self {
            Token::Block { content, .. }
            | Token::Comment { content, .. }
            | Token::Error { content, .. }
            | Token::Text { content, .. }
            | Token::Variable { content, .. } => content.clone(),
            Token::Whitespace { span, .. } => " ".repeat(span.length_usize()),
            Token::Newline { span, .. } => {
                if span.length() == 2 {
                    "\r\n".to_string()
                } else {
                    "\n".to_string()
                }
            }
            Token::Eof => String::new(),
        }
    }

    /// Get the lexeme as it appears in source
    #[must_use]
    pub fn lexeme(&self) -> String {
        match self {
            Token::Block { content, .. } => format!(
                "{} {} {}",
                TagDelimiter::Block.opener(),
                content,
                TagDelimiter::Block.closer()
            ),
            Token::Variable { content, .. } => format!(
                "{} {} {}",
                TagDelimiter::Variable.opener(),
                content,
                TagDelimiter::Variable.closer()
            ),
            Token::Comment { content, .. } => format!(
                "{} {} {}",
                TagDelimiter::Comment.opener(),
                content,
                TagDelimiter::Comment.closer()
            ),
            Token::Text { content, .. } | Token::Error { content, .. } => content.clone(),
            Token::Whitespace { span, .. } => " ".repeat(span.length_usize()),
            Token::Newline { span, .. } => {
                if span.length() == 2 {
                    "\r\n".to_string()
                } else {
                    "\n".to_string()
                }
            }
            Token::Eof => String::new(),
        }
    }

    #[must_use]
    pub fn offset(&self) -> Option<u32> {
        match self {
            Token::Block { span, .. }
            | Token::Comment { span, .. }
            | Token::Error { span, .. }
            | Token::Variable { span, .. } => {
                Some(span.start().saturating_sub(TagDelimiter::LENGTH_U32))
            }
            Token::Text { span, .. }
            | Token::Whitespace { span, .. }
            | Token::Newline { span, .. } => Some(span.start()),
            Token::Eof => None,
        }
    }

    /// Get the length of the token content
    #[must_use]
    pub fn length(&self) -> u32 {
        let len = match self {
            Token::Block { content, .. }
            | Token::Comment { content, .. }
            | Token::Error { content, .. }
            | Token::Text { content, .. }
            | Token::Variable { content, .. } => content.len(),
            Token::Whitespace { span, .. } | Token::Newline { span, .. } => span.length_usize(),
            Token::Eof => 0,
        };
        u32::try_from(len).unwrap_or(u32::MAX)
    }

    #[must_use]
    pub fn full_span(&self) -> Option<Span> {
        match self {
            Token::Block { span, .. }
            | Token::Comment { span, .. }
            | Token::Variable { span, .. } => {
                Some(span.expand(TagDelimiter::LENGTH_U32, TagDelimiter::LENGTH_U32))
            }
            Token::Error { span, .. } => Some(span.expand(TagDelimiter::LENGTH_U32, 0)),
            Token::Newline { span, .. }
            | Token::Text { span, .. }
            | Token::Whitespace { span, .. } => Some(*span),
            Token::Eof => None,
        }
    }

    #[must_use]
    pub fn content_span(&self) -> Option<Span> {
        match self {
            Token::Block { span, .. }
            | Token::Comment { span, .. }
            | Token::Error { span, .. }
            | Token::Text { span, .. }
            | Token::Variable { span, .. }
            | Token::Whitespace { span, .. }
            | Token::Newline { span, .. } => Some(*span),
            Token::Eof => None,
        }
    }

    #[must_use]
    pub fn full_span_or_fallback(&self) -> Span {
        self.full_span()
            .unwrap_or_else(|| self.content_span_or_fallback())
    }

    #[must_use]
    pub fn content_span_or_fallback(&self) -> Span {
        self.content_span()
            .unwrap_or_else(|| Span::new(self.offset().unwrap_or(0), self.length()))
    }

    #[must_use]
    pub fn spans(&self) -> (Span, Span) {
        let content = self.content_span_or_fallback();
        let full = self.full_span().unwrap_or(content);
        (content, full)
    }
}

#[cfg(test)]
#[derive(Debug, serde::Serialize)]
pub enum TokenSnapshot {
    Block {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Comment {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Eof,
    Error {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Newline {
        span: (u32, u32),
    },
    Text {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Variable {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Whitespace {
        span: (u32, u32),
    },
}

#[cfg(test)]
impl Token {
    /// ## Panics
    ///
    /// This may panic on the `full_span` calls, but it's only used in testing,
    /// so it's all good.
    #[must_use]
    pub fn to_snapshot(&self) -> TokenSnapshot {
        match self {
            Token::Block { span, .. } => TokenSnapshot::Block {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Comment { span, .. } => TokenSnapshot::Comment {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Eof => TokenSnapshot::Eof,
            Token::Error { span, .. } => TokenSnapshot::Error {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Newline { span } => TokenSnapshot::Newline { span: span.into() },
            Token::Text { span, .. } => TokenSnapshot::Text {
                content: self.content(),
                span: span.into(),
                full_span: span.into(),
            },
            Token::Variable { span, .. } => TokenSnapshot::Variable {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Whitespace { span } => TokenSnapshot::Whitespace { span: span.into() },
        }
    }
}

#[cfg(test)]
pub struct TokenSnapshotVec(pub Vec<Token>);

#[cfg(test)]
impl TokenSnapshotVec {
    #[must_use]
    pub fn to_snapshot(&self) -> Vec<TokenSnapshot> {
        self.0.iter().map(Token::to_snapshot).collect()
    }
}

#[derive(Debug, Clone)]
pub struct TokenStream(Vec<Token>);

impl TokenStream {
    const CHARS_PER_TOKEN: usize = 6;
    const MIN_CAPACITY: usize = 32;
    const MAX_CAPACITY: usize = 1024;

    #[must_use]
    pub fn with_estimated_capacity(source: &str) -> Self {
        let capacity =
            (source.len() / Self::CHARS_PER_TOKEN).clamp(Self::MIN_CAPACITY, Self::MAX_CAPACITY);
        Self(Vec::with_capacity(capacity))
    }

    #[inline]
    pub fn push(&mut self, token: Token) {
        self.0.push(token);
    }

    /// Get the number of tokens in the stream.
    #[must_use]
    pub fn len(&self) -> usize {
        self.0.len()
    }

    /// Get the number of content tokens (excluding EOF).
    #[must_use]
    pub fn content_len(&self) -> usize {
        self.0.len().saturating_sub(1)
    }

    /// Check if stream is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

impl From<TokenStream> for Vec<Token> {
    fn from(val: TokenStream) -> Self {
        val.0
    }
}

impl IntoIterator for TokenStream {
    type Item = Token;
    type IntoIter = std::vec::IntoIter<Token>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}
