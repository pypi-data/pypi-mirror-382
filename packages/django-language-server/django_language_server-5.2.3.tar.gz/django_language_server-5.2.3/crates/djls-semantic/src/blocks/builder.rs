use djls_source::Span;
use djls_templates::tokens::TagDelimiter;
use djls_templates::Node;
use salsa::Accumulator;

use super::grammar::CloseValidation;
use super::grammar::TagClass;
use super::grammar::TagIndex;
use super::tree::BlockId;
use super::tree::BlockNode;
use super::tree::BlockTree;
use super::tree::Blocks;
use super::tree::BranchKind;
use crate::traits::SemanticModel;
use crate::Db;
use crate::ValidationError;
use crate::ValidationErrorAccumulator;

#[derive(Debug, Clone)]
enum TreeOp {
    AddRoot {
        id: BlockId,
    },
    AddBranchNode {
        target: BlockId,
        tag: String,
        marker_span: Span,
        body: BlockId,
        kind: BranchKind,
    },
    AddLeafNode {
        target: BlockId,
        label: String,
        span: Span,
    },
    ExtendBlockSpan {
        id: BlockId,
        span: Span,
    },
    FinalizeSpanTo {
        id: BlockId,
        end: u32,
    },
    AccumulateDiagnostic(ValidationError),
}

pub struct BlockTreeBuilder<'db> {
    db: &'db dyn Db,
    index: TagIndex<'db>,
    stack: Vec<TreeFrame>,
    block_allocs: Vec<(Span, Option<BlockId>)>,
    ops: Vec<TreeOp>,
}

impl<'db> BlockTreeBuilder<'db> {
    #[allow(dead_code)] // use is gated behind cfg(test) for now
    pub fn new(db: &'db dyn Db, index: TagIndex<'db>) -> Self {
        Self {
            db,
            index,
            stack: Vec::new(),
            block_allocs: Vec::new(),
            ops: Vec::new(),
        }
    }

    /// Allocate a new `BlockId` and track its metadata for later creation
    fn alloc_block_id(&mut self, span: Span, parent: Option<BlockId>) -> BlockId {
        let id = BlockId::new(u32::try_from(self.block_allocs.len()).unwrap_or_default());
        self.block_allocs.push((span, parent));
        id
    }

    /// Apply all semantic operations to build a tracked `BlockTree`
    fn apply_operations(self) -> BlockTree<'db> {
        let BlockTreeBuilder {
            db,
            block_allocs,
            ops,
            ..
        } = self;

        let mut roots = Vec::new();
        let mut blocks = Blocks::default();

        for (span, parent) in block_allocs {
            if let Some(p) = parent {
                blocks.alloc(span, Some(p));
            } else {
                blocks.alloc(span, None);
            }
        }

        for op in ops {
            match op {
                TreeOp::AddRoot { id } => {
                    roots.push(id);
                }
                TreeOp::AddBranchNode {
                    target,
                    tag,
                    marker_span,
                    body,
                    kind,
                } => {
                    blocks.push_node(
                        target,
                        BlockNode::Branch {
                            tag,
                            marker_span,
                            body,
                            kind,
                        },
                    );
                }
                TreeOp::AddLeafNode {
                    target,
                    label,
                    span,
                } => {
                    blocks.push_node(target, BlockNode::Leaf { label, span });
                }
                TreeOp::ExtendBlockSpan { id, span } => {
                    blocks.extend_block(id, span);
                }
                TreeOp::FinalizeSpanTo { id, end } => {
                    blocks.finalize_block_span(id, end);
                }
                TreeOp::AccumulateDiagnostic(error) => {
                    ValidationErrorAccumulator(error).accumulate(db);
                }
            }
        }

        BlockTree::new(db, roots, blocks)
    }

    fn handle_tag(&mut self, name: &str, bits: &[String], span: Span) {
        let full_span = expand_marker(span);
        match self.index.classify(self.db, name) {
            TagClass::Opener => {
                let parent = get_active_segment(&self.stack);

                let container = self.alloc_block_id(span, parent);
                let segment = self.alloc_block_id(
                    Span::new(span.end().saturating_add(TagDelimiter::LENGTH_U32), 0),
                    Some(container),
                );

                if let Some(parent_id) = parent {
                    // Nested block
                    self.ops.push(TreeOp::AddBranchNode {
                        target: parent_id,
                        tag: name.to_string(),
                        marker_span: span,
                        body: container,
                        kind: BranchKind::Opener,
                    });
                    self.ops.push(TreeOp::AddBranchNode {
                        target: container,
                        tag: name.to_string(),
                        marker_span: span,
                        body: segment,
                        kind: BranchKind::Segment,
                    });
                } else {
                    // Root block
                    self.ops.push(TreeOp::AddRoot { id: container });
                    self.ops.push(TreeOp::AddBranchNode {
                        target: container,
                        tag: name.to_string(),
                        marker_span: span,
                        body: segment,
                        kind: BranchKind::Segment,
                    });
                }

                self.stack.push(TreeFrame {
                    opener_name: name.to_string(),
                    opener_bits: bits.to_vec(),
                    opener_span: full_span,
                    container_body: container,
                    segment_body: segment,
                });
            }
            TagClass::Closer { opener_name } => {
                self.close_block(&opener_name, bits, span);
            }
            TagClass::Intermediate { possible_openers } => {
                self.add_intermediate(name, &possible_openers, span);
            }
            TagClass::Unknown => {
                if let Some(segment) = get_active_segment(&self.stack) {
                    self.ops.push(TreeOp::AddLeafNode {
                        target: segment,
                        label: name.to_string(),
                        span,
                    });
                }
            }
        }
    }

    fn close_block(&mut self, opener_name: &str, closer_bits: &[String], span: Span) {
        let marker_span = expand_marker(span);

        let Some(frame_idx) = find_frame_from_opener(&self.stack, opener_name) else {
            self.ops.push(TreeOp::AccumulateDiagnostic(
                ValidationError::UnbalancedStructure {
                    opening_tag: opener_name.to_string(),
                    expected_closing: String::new(),
                    opening_span: marker_span,
                    closing_span: None,
                },
            ));
            return;
        };

        // Pop any unclosed blocks above this one
        while self.stack.len() > frame_idx + 1 {
            if let Some(unclosed) = self.stack.pop() {
                self.ops
                    .push(TreeOp::AccumulateDiagnostic(ValidationError::UnclosedTag {
                        tag: unclosed.opener_name,
                        span: unclosed.opener_span,
                    }));
            }
        }

        // validate and close
        let frame = self.stack.pop().unwrap();
        match self
            .index
            .validate_close(self.db, opener_name, &frame.opener_bits, closer_bits)
        {
            CloseValidation::Valid => {
                // Finalize the last segment body to end just before the closer marker
                let content_end = span.start().saturating_sub(TagDelimiter::LENGTH_U32);
                self.ops.push(TreeOp::FinalizeSpanTo {
                    id: frame.segment_body,
                    end: content_end,
                });
                // Extend to include closer
                self.ops.push(TreeOp::ExtendBlockSpan {
                    id: frame.container_body,
                    span,
                });
            }
            CloseValidation::ArgumentMismatch { expected, got, .. } => {
                let name = if got.is_empty() { expected } else { got };
                self.ops.push(TreeOp::AccumulateDiagnostic(
                    ValidationError::UnmatchedBlockName {
                        name,
                        span: marker_span,
                    },
                ));
                self.ops
                    .push(TreeOp::AccumulateDiagnostic(ValidationError::UnclosedTag {
                        tag: frame.opener_name.clone(),
                        span: frame.opener_span,
                    }));
                self.stack.push(frame); // Restore frame
            }
            CloseValidation::MissingRequiredArg { expected, .. } => {
                let expected_closing = format!("{} {}", frame.opener_name, expected);
                self.ops.push(TreeOp::AccumulateDiagnostic(
                    ValidationError::UnbalancedStructure {
                        opening_tag: frame.opener_name.clone(),
                        expected_closing,
                        opening_span: frame.opener_span,
                        closing_span: Some(marker_span),
                    },
                ));
                self.stack.push(frame);
            }
            CloseValidation::UnexpectedArg { arg, got } => {
                let name = if got.is_empty() { arg } else { got };
                self.ops.push(TreeOp::AccumulateDiagnostic(
                    ValidationError::UnmatchedBlockName {
                        name,
                        span: marker_span,
                    },
                ));
                self.ops
                    .push(TreeOp::AccumulateDiagnostic(ValidationError::UnclosedTag {
                        tag: frame.opener_name.clone(),
                        span: frame.opener_span,
                    }));
                self.stack.push(frame);
            }
            CloseValidation::NotABlock => {
                self.ops.push(TreeOp::AccumulateDiagnostic(
                    ValidationError::UnbalancedStructure {
                        opening_tag: opener_name.to_string(),
                        expected_closing: opener_name.to_string(),
                        opening_span: frame.opener_span,
                        closing_span: Some(marker_span),
                    },
                ));
                self.stack.push(frame);
            }
        }
    }

    fn add_intermediate(&mut self, tag_name: &str, possible_openers: &[String], span: Span) {
        let marker_span = expand_marker(span);

        if let Some(frame) = self.stack.last() {
            if possible_openers.contains(&frame.opener_name) {
                // Finalize previous segment body to just before this marker (full start)
                let content_end = span.start().saturating_sub(TagDelimiter::LENGTH_U32);
                let segment_to_finalize = frame.segment_body;
                let container = frame.container_body;

                self.ops.push(TreeOp::FinalizeSpanTo {
                    id: segment_to_finalize,
                    end: content_end,
                });

                let body_start = span.end().saturating_add(TagDelimiter::LENGTH_U32);
                let new_segment_id = self.alloc_block_id(Span::new(body_start, 0), Some(container));

                // Add the branch node for the new segment
                self.ops.push(TreeOp::AddBranchNode {
                    target: container,
                    tag: tag_name.to_string(),
                    marker_span: span,
                    body: new_segment_id,
                    kind: BranchKind::Segment,
                });

                self.stack.last_mut().unwrap().segment_body = new_segment_id;
            } else {
                let context = format_intermediate_context(possible_openers);
                self.ops
                    .push(TreeOp::AccumulateDiagnostic(ValidationError::OrphanedTag {
                        tag: tag_name.to_string(),
                        context,
                        span: marker_span,
                    }));
            }
        } else {
            let context = format_intermediate_context(possible_openers);
            self.ops
                .push(TreeOp::AccumulateDiagnostic(ValidationError::OrphanedTag {
                    tag: tag_name.to_string(),
                    context,
                    span: marker_span,
                }));
        }
    }

    fn finish(&mut self) {
        while let Some(frame) = self.stack.pop() {
            if self.index.is_end_optional(self.db, &frame.opener_name) {
                // No explicit closer: finalize last segment to end of input (best-effort)
                // We do not know the real end; leave as-is and extend container by opener span only.
                self.ops.push(TreeOp::ExtendBlockSpan {
                    id: frame.container_body,
                    span: frame.opener_span,
                });
            } else {
                self.ops
                    .push(TreeOp::AccumulateDiagnostic(ValidationError::UnclosedTag {
                        tag: frame.opener_name,
                        span: frame.opener_span,
                    }));
            }
        }
    }
}

fn expand_marker(span: Span) -> Span {
    span.expand(TagDelimiter::LENGTH_U32, TagDelimiter::LENGTH_U32)
}

fn format_intermediate_context(possible_openers: &[String]) -> String {
    match possible_openers.len() {
        0 => "a valid parent block".to_string(),
        1 => format!("'{}' block", possible_openers[0]),
        2 => format!(
            "'{}' or '{}' block",
            possible_openers[0], possible_openers[1]
        ),
        _ => {
            let mut parts = possible_openers
                .iter()
                .map(|name| format!("'{name}'"))
                .collect::<Vec<_>>();
            let last = parts.pop().unwrap_or_default();
            let prefix = parts.join(", ");
            format!("one of {prefix}, or {last} blocks")
        }
    }
}

type TreeStack = Vec<TreeFrame>;

/// Get the currently active segment (the innermost block we're in)
fn get_active_segment(stack: &TreeStack) -> Option<BlockId> {
    stack.last().map(|frame| frame.segment_body)
}

/// Find a frame in the stack by name
fn find_frame_from_opener(stack: &TreeStack, opener_name: &str) -> Option<usize> {
    stack.iter().rposition(|f| f.opener_name == opener_name)
}

struct TreeFrame {
    opener_name: String,
    opener_bits: Vec<String>,
    opener_span: Span,
    container_body: BlockId,
    segment_body: BlockId,
}

impl<'db> SemanticModel<'db> for BlockTreeBuilder<'db> {
    type Model = BlockTree<'db>;

    fn observe(&mut self, node: Node) {
        match node {
            Node::Tag { name, bits, span } => {
                self.handle_tag(&name, &bits, span);
            }
            Node::Comment { span, .. } => {
                if let Some(parent) = get_active_segment(&self.stack) {
                    self.ops.push(TreeOp::AddLeafNode {
                        target: parent,
                        label: "<comment>".into(),
                        span,
                    });
                }
            }
            Node::Variable { span, .. } => {
                if let Some(parent) = get_active_segment(&self.stack) {
                    self.ops.push(TreeOp::AddLeafNode {
                        target: parent,
                        label: "<var>".into(),
                        span,
                    });
                }
            }
            Node::Error {
                full_span, error, ..
            } => {
                if let Some(parent) = get_active_segment(&self.stack) {
                    self.ops.push(TreeOp::AddLeafNode {
                        target: parent,
                        label: error.to_string(),
                        span: full_span,
                    });
                }
            }
            Node::Text { .. } => {} // Skip text nodes - we only care about Django constructs
        }
    }

    fn construct(mut self) -> Self::Model {
        self.finish();
        self.apply_operations()
    }
}
