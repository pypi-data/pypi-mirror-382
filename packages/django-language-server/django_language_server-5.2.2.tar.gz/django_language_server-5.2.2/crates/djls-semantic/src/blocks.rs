mod builder;
mod grammar;
mod snapshot;
mod tree;

use builder::BlockTreeBuilder;
pub use grammar::TagIndex;
pub(crate) use tree::BlockId;
pub(crate) use tree::BlockNode;
pub(crate) use tree::BlockTree;
pub(crate) use tree::BranchKind;

use crate::db::Db;
use crate::traits::SemanticModel;

#[salsa::tracked]
pub fn build_block_tree<'db>(
    db: &'db dyn Db,
    nodelist: djls_templates::NodeList<'db>,
) -> BlockTree<'db> {
    let builder = BlockTreeBuilder::new(db, db.tag_index());
    builder.model(db, nodelist)
}
