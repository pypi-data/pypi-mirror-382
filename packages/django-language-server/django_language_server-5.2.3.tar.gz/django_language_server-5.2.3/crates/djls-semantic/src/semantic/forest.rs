use djls_source::Span;
use djls_templates::tokens::TagDelimiter;
use djls_templates::Node;
use rustc_hash::FxHashSet;
use serde::Serialize;

use crate::blocks::BlockId;
use crate::blocks::BlockNode;
use crate::blocks::BlockTree;
use crate::blocks::BranchKind;
use crate::Db;

#[salsa::tracked]
pub struct SemanticForest<'db> {
    #[returns(ref)]
    pub roots: Vec<SemanticNode>,
    #[returns(ref)]
    pub tag_spans: Vec<Span>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
pub enum SemanticNode {
    Tag {
        name: String,
        marker_span: Span,
        arguments: Vec<String>,
        segments: Vec<SemanticSegment>,
    },
    Leaf {
        label: String,
        span: Span,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
pub struct SemanticSegment {
    pub kind: SegmentKind,
    pub marker_span: Span,
    pub content_span: Span,
    pub arguments: Vec<String>,
    pub children: Vec<SemanticNode>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
pub enum SegmentKind {
    Main,
    Intermediate { tag: String },
}

pub fn build_root_tag(
    db: &dyn Db,
    tree: BlockTree,
    nodelist: djls_templates::NodeList<'_>,
    container_id: BlockId,
    spans: &mut FxHashSet<Span>,
) -> Option<SemanticNode> {
    let container = tree.blocks(db).get(container_id.index());
    for node in container.nodes() {
        if let BlockNode::Branch {
            tag,
            marker_span,
            kind: BranchKind::Segment,
            ..
        } = node
        {
            spans.insert(expand_marker(*marker_span));
            return Some(build_tag_from_container(
                db,
                tree,
                nodelist,
                container_id,
                tag.clone(),
                *marker_span,
                spans,
            ));
        }
    }
    None
}

fn build_tag_from_container(
    db: &dyn Db,
    tree: BlockTree,
    nodelist: djls_templates::NodeList<'_>,
    container_id: BlockId,
    tag_name: String,
    opener_marker_span: Span,
    spans: &mut FxHashSet<Span>,
) -> SemanticNode {
    let segments = build_segments(db, tree, nodelist, container_id, opener_marker_span, spans);
    let arguments = segments
        .first()
        .map(|segment| segment.arguments.clone())
        .unwrap_or_default();

    SemanticNode::Tag {
        name: tag_name,
        marker_span: opener_marker_span,
        arguments,
        segments,
    }
}

fn build_segments(
    db: &dyn Db,
    tree: BlockTree,
    nodelist: djls_templates::NodeList<'_>,
    container_id: BlockId,
    opener_marker_span: Span,
    spans: &mut FxHashSet<Span>,
) -> Vec<SemanticSegment> {
    let mut segments = Vec::new();
    let container = tree.blocks(db).get(container_id.index());

    for (idx, node) in container.nodes().iter().enumerate() {
        if let BlockNode::Branch {
            tag,
            marker_span,
            body,
            kind: BranchKind::Segment,
        } = node
        {
            let kind = if idx == 0 {
                SegmentKind::Main
            } else {
                SegmentKind::Intermediate { tag: tag.clone() }
            };

            let marker = if idx == 0 {
                opener_marker_span
            } else {
                *marker_span
            };

            spans.insert(expand_marker(marker));

            let content_block = tree.blocks(db).get(body.index());
            let arguments = lookup_arguments(db, nodelist, marker);
            let children = build_children(db, tree, nodelist, *body, spans);

            segments.push(SemanticSegment {
                kind,
                marker_span: marker,
                content_span: *content_block.span(),
                arguments,
                children,
            });
        }
    }

    segments
}

fn build_children(
    db: &dyn Db,
    tree: BlockTree,
    nodelist: djls_templates::NodeList<'_>,
    block_id: BlockId,
    spans: &mut FxHashSet<Span>,
) -> Vec<SemanticNode> {
    let mut children = Vec::new();
    let block = tree.blocks(db).get(block_id.index());

    for node in block.nodes() {
        match node {
            BlockNode::Leaf { label, span } => {
                children.push(SemanticNode::Leaf {
                    label: label.clone(),
                    span: *span,
                });
            }
            BlockNode::Branch {
                tag,
                marker_span,
                body,
                kind: BranchKind::Opener | BranchKind::Segment,
            } => {
                spans.insert(expand_marker(*marker_span));
                children.push(build_tag_from_container(
                    db,
                    tree,
                    nodelist,
                    *body,
                    tag.clone(),
                    *marker_span,
                    spans,
                ));
            }
        }
    }

    children
}

fn lookup_arguments(
    db: &dyn Db,
    nodelist: djls_templates::NodeList<'_>,
    marker_span: Span,
) -> Vec<String> {
    nodelist
        .nodelist(db)
        .iter()
        .find_map(|node| match node {
            Node::Tag { bits, span, .. } if *span == marker_span => Some(bits.clone()),
            _ => None,
        })
        .unwrap_or_default()
}

fn expand_marker(span: Span) -> Span {
    span.expand(TagDelimiter::LENGTH_U32, TagDelimiter::LENGTH_U32)
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;
    use std::sync::Mutex;

    use camino::Utf8Path;
    use camino::Utf8PathBuf;
    use djls_source::File;
    use djls_templates::parse_template;
    use djls_workspace::FileSystem;
    use djls_workspace::InMemoryFileSystem;
    use insta::assert_yaml_snapshot;

    use super::*;
    use crate::blocks::build_block_tree;
    use crate::build_semantic_forest;
    use crate::templatetags::django_builtin_specs;
    use crate::TagIndex;

    #[salsa::db]
    #[derive(Clone)]
    struct TestDatabase {
        storage: salsa::Storage<Self>,
        fs: Arc<Mutex<InMemoryFileSystem>>,
    }

    impl TestDatabase {
        fn new() -> Self {
            Self {
                storage: salsa::Storage::default(),
                fs: Arc::new(Mutex::new(InMemoryFileSystem::new())),
            }
        }

        fn add_file(&self, path: &str, content: &str) {
            self.fs
                .lock()
                .unwrap()
                .add_file(path.into(), content.to_string());
        }
    }

    #[salsa::db]
    impl salsa::Database for TestDatabase {}

    #[salsa::db]
    impl djls_source::Db for TestDatabase {
        fn create_file(&self, path: &Utf8Path) -> File {
            File::new(self, path.to_owned(), 0)
        }

        fn get_file(&self, _path: &Utf8Path) -> Option<File> {
            None
        }

        fn read_file(&self, path: &Utf8Path) -> std::io::Result<String> {
            self.fs.lock().unwrap().read_to_string(path)
        }
    }

    #[salsa::db]
    impl djls_templates::Db for TestDatabase {}

    #[salsa::db]
    impl crate::Db for TestDatabase {
        fn tag_specs(&self) -> crate::templatetags::TagSpecs {
            django_builtin_specs()
        }

        fn tag_index(&self) -> TagIndex<'_> {
            TagIndex::from_specs(self)
        }

        fn template_dirs(&self) -> Option<Vec<Utf8PathBuf>> {
            None
        }
    }

    #[test]
    fn semantic_forest_snapshot() {
        let db = TestDatabase::new();
        let source = r"
{% block header %}
    <h1>Title</h1>
{% endblock header %}

{% if user.is_authenticated %}
    {% for item in items %}
        <li>{{ item }}</li>
    {% empty %}
        <li>No items</li>
    {% endfor %}
{% else %}
    <p>Please log in</p>
{% endif %}
";

        db.add_file("template.html", source);
        let file = File::new(&db, "template.html".into(), 0);
        let nodelist = parse_template(&db, file).expect("should parse");

        let block_tree = build_block_tree(&db, nodelist);
        let forest = build_semantic_forest(&db, block_tree, nodelist);

        assert_yaml_snapshot!(ForestSnapshot::capture(forest, &db));
    }

    #[test]
    fn semantic_forest_intermediate_snapshot() {
        let db = TestDatabase::new();
        let source = r"
{% if user.is_staff %}
    <span>Staff</span>
{% elif user.is_manager %}
    <span>Manager</span>
{% else %}
    <span>Regular</span>
{% endif %}
";

        db.add_file("intermediate.html", source);
        let file = File::new(&db, "intermediate.html".into(), 0);
        let nodelist = parse_template(&db, file).expect("should parse");

        let block_tree = build_block_tree(&db, nodelist);
        let forest = build_semantic_forest(&db, block_tree, nodelist);

        assert_yaml_snapshot!("intermediate", ForestSnapshot::capture(forest, &db));
    }

    #[derive(Serialize)]
    struct ForestSnapshot {
        roots: Vec<SemanticNode>,
        tag_spans: Vec<Span>,
    }

    impl ForestSnapshot {
        fn capture(forest: SemanticForest, db: &dyn crate::Db) -> Self {
            let mut tag_spans = forest.tag_spans(db).clone();
            tag_spans.sort_by_key(|span| (span.start(), span.end()));

            Self {
                roots: forest.roots(db).clone(),
                tag_spans,
            }
        }
    }
}
