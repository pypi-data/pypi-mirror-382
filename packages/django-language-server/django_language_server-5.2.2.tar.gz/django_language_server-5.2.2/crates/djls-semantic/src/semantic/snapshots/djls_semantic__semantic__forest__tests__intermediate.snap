---
source: crates/djls-semantic/src/semantic/forest.rs
expression: "&forest"
---
roots:
  - Tag:
      name: if
      marker_span:
        start: 3
        length: 18
      arguments:
        - user.is_staff
      segments:
        - kind: Main
          marker_span:
            start: 3
            length: 18
          content_span:
            start: 23
            length: 24
          arguments:
            - user.is_staff
          children: []
        - kind:
            Intermediate:
              tag: elif
          marker_span:
            start: 49
            length: 22
          content_span:
            start: 73
            length: 26
          arguments:
            - user.is_manager
          children: []
        - kind:
            Intermediate:
              tag: else
          marker_span:
            start: 101
            length: 6
          content_span:
            start: 109
            length: 26
          arguments: []
          children: []
tag_spans:
  - start: 1
    length: 22
  - start: 47
    length: 26
  - start: 99
    length: 10
